### PR TITLE
Test coupling heat transfer/two phase flow

### PR DESCRIPTION
**Pull request by mschreter**
In this PR the coupled heat transfer two-phase flow problem will be tested for the first time. To this end, a thermo capillary droplet migration according to [Nas, Tryggvason 2003](https://www.sciencedirect.com/science/article/pii/S0301932203000843?casa_token=Sw9xFN3LLFIAAAAA:6U4NSOa_hNtNYYvYM8d4ctMo_M6XfzqLOYRFTfctMUa5cmOLFKXPisJpgGqvB2fu66AnxL3yWyE) will be simulated. 

 

Results for the velocity field from Mah, Bothe 2011:
 

Our results:
 

@nmuch: This is a first verification example for the thermo-fluid coupling. Would you like to add another one (maybe the two bubbles) from the publication of Nas, Tryggvason 2003?

### CODE REVIEW

**Comment by peterrum** [comment]
Nice progress :+1:

**Comment by nmuch** [comment]
Very nice! 
Sure I'll add the two bubble example. I'll do it in another PR.

For me the `tests/thermo_capillary_droplet.json` runs but the `include/meltpooldg/simulations/thermo_capillary_droplet/thermo_capillary_droplet.json` does not. I get this error message:
[error.txt](https://github.com/MeltPoolDG/MeltPoolDG/files/6221121/error.txt).
I guess the paraview output does not work correctly yet. Can you look into this?

**Comment by mschreter** [comment]
> Very nice!
> Sure I'll add the two bubble example. I'll do it in another PR.
> 
> For me the `tests/thermo_capillary_droplet.json` runs but the `include/meltpooldg/simulations/thermo_capillary_droplet/thermo_capillary_droplet.json` does not. I get this error message:
> [error.txt](https://github.com/MeltPoolDG/MeltPoolDG/files/6221121/error.txt).
> I guess the paraview output does not work correctly yet. Can you look into this?

Cool, thanks :-). Ah, here the folder "output_thermo_capillary" must exist in advance to the simulation. We could force MeltPoolDG to automatically create the folder if it does not exist. I will fix this problem in a follow-up PR.

File: include/meltpooldg/heat/heat_transfer_operation.hpp [hunk 1]
```diff
@@ -180,27 +180,6 @@ namespace MeltPoolDG::Heat
...
-                                   std::vector<std::string>(dim, "velocity"),
-                                   vector_component_interpretation);
-        }
-
-      if (level_set_as_heaviside)
-        {
-          data_out.add_data_vector(scratch_data.get_dof_handler(ls_dof_idx),
-                                   *level_set_as_heaviside,
-                                   "level_set_as_heaviside");
-        }
```
Comment (peterrum) [inline include/meltpooldg/heat/heat_transfer_operation.hpp:203]: Why did you remove this?

File: include/meltpooldg/flow/surface_tension_operation.hpp [hunk 2]
```diff
@@ -14,6 +14,26 @@ namespace MeltPoolDG::Flow
...
+   *  to temperature-(in)dependent surface tension forces
+   *
+   *          /                \        /                         \
+   *  f_st =  | N_a,  α κ n  δ |   +    | N_a,  (I - n ⊗ n ) ∇α δ |
+   *          \            Γ   /        \             Γ   Γ       /
+   *                            Ω                                  Ω
+   *
+   *  with the temperature-dependent surface tension
+   *
+   *  α = α_0 - α'_0 ( T - T    )
```
Comment (peterrum) [inline include/meltpooldg/flow/surface_tension_operation.hpp:28]:
```suggestion
   *  α = min(α_0, α_0 - α'_0 ( T - T    ))
```
as far as I understand the code.

File: include/meltpooldg/flow/surface_tension_operation.hpp [hunk 3]
```diff
@@ -151,22 +187,22 @@ namespace MeltPoolDG::Flow
...
 
                   auto alpha = compare_and_apply_mask<SIMDComparison::less_than>(
                     T,
                     T0,
                     VectorizedArray<double>(alpha0),
                     VectorizedArray<double>(alpha0) - VectorizedArray<double>(d_alpha0) * (T - T0));
 
+                  for (unsigned int v = 0; v < matrix_free.n_active_entries_per_cell_batch(cell);
+                       ++v)
+                    alpha[v] = alpha[v] < 1e-10 ? 0.0 : alpha[v];
```
Comment (peterrum) [inline include/meltpooldg/flow/surface_tension_operation.hpp:202]: Why aren't you use `compare_and_apply_mask` here as well?

**Review by peterrum — APPROVED** [review]
Just some minor questions and remarks from my side.

File: include/meltpooldg/heat/heat_transfer_operation.hpp (hunk 1 above)
Comment (mschreter) [reply include/meltpooldg/heat/heat_transfer_operation.hpp:203]: When I called `heat_transfer_operation->attach_output_vectors` from `two_phase_flow_problem.hpp`, then I got an assert from `DataOut` that vectors with names "velocity" and "level_set_as_heaviside" are already registered (namely from `flow_operation` and `level_set_operation`). Of course, when we run now `heat_transfer_problem` with a prescribed `velocity` or `level_set_as_heaviside`, we would not get the corresponding paraview output anymore. However, imho these cases serve primarily for testing and thus, I would consider its complete paraview output not extremely important. If you agree, I would simply remove them here to avoid further `DataOut` conflicts. @nmuch: What is your opinion here?

Comment (nmuch) [reply include/meltpooldg/heat/heat_transfer_operation.hpp:203]: I agree. I added these outputs when creating tests for heat transfer with (two-phase) velocity to have a look at these fields. It makes more sense to let the two-phase flow problem output these fields.

File: include/meltpooldg/heat/heat_transfer_operator.hpp [hunk 4]
```diff
@@ -162,7 +162,7 @@ namespace MeltPoolDG::Heat
     {
       AssertThrow(this->d_tau > 0.0, ExcMessage("advection diffusion operator: d_tau must be set"));
 
-      MeltPoolDG::VectorTools::update_ghost_values(temperature);
+      MeltPoolDG::VectorTools::update_ghost_values(temperature, heat_source);
```
[the remark this answers is not in the document]
Comment (mschreter) [reply include/meltpooldg/heat/heat_transfer_operator.hpp:165]: Yes. But it has not been noticed since we have not applied a heat_source up to now.

File: include/meltpooldg/heat/heat_transfer_operator.hpp [hunk 5]
```diff
@@ -449,6 +449,10 @@ namespace MeltPoolDG::Heat
     create_rhs(VectorType &dst, const VectorType &src /*temperature_old*/) const override
     {
       MeltPoolDG::VectorTools::update_ghost_values(temperature, heat_source);
+      if (velocity)
+        MeltPoolDG::VectorTools::update_ghost_values(*velocity);
+      if (level_set_as_heaviside)
+        MeltPoolDG::VectorTools::update_ghost_values(*level_set_as_heaviside);
```
[the remark this answers is not in the document]
Comment (mschreter) [reply include/meltpooldg/heat/heat_transfer_operator.hpp:455]: Yes, they were missing.

File: include/meltpooldg/flow/surface_tension_operation.hpp (hunk 2 above)
Comment (mschreter) [reply include/meltpooldg/flow/surface_tension_operation.hpp:28]: You're right. I've added a note on that.

Comment (mschreter) [reply include/meltpooldg/flow/surface_tension_operation.hpp:28]: I've clarified this issue.

File: include/meltpooldg/flow/surface_tension_operation.hpp (hunk 3 above)
Comment (mschreter) [reply include/meltpooldg/flow/surface_tension_operation.hpp:202]: Done.

**Review by peterrum — APPROVED** [review]
